File: wander/src/test/kotlin/dev/ligature/wander/WanderSuite.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.wander

import arrow.core.Either
import dev.ligature.IntegerLiteral
import dev.ligature.inmemory.InMemoryLigature
import dev.ligature.wander.interpreter.IntegerPrimitive
import dev.ligature.wander.writer.Writer
import io.kotest.core.spec.style.FunSpec
import io.kotest.inspectors.forAll
import io.kotest.matchers.shouldBe
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.io.path.ExperimentalPathApi
import kotlin.io.path.extension
import kotlin.io.path.readText
import kotlin.streams.toList

@OptIn(ExperimentalPathApi::class)
class WanderSuite : FunSpec() {
    private val ligature = InMemoryLigature()
    private val wander = Wander(ligature)
    private val writer = Writer()

    private fun readDirectory(path: String): List<Path> =
        Files.walk(Paths.get(path)).toList()

    data class TestResult(val expected: String, val commandResult: String, val queryResult: String)

    private fun buildResults(path: String): List<TestResult> {
        val files = readDirectory(path)
        return files.filter { it.extension == "wander" }.map {
            val text = it.readText()
            val queryResult = writer.write(wander.runQuery(text))
            val commandResult = writer.write(wander.runCommand(text))
            val expected = it.resolveSibling(it.fileName.toString().replace(".wander", ".result")).readText()
            TestResult(expected, commandResult, queryResult)
        }
    }

    init {
        test("integer support") {
            val script = "5"
            val res = wander.run(script)
            res shouldBe Either.Right(IntegerPrimitive(IntegerLiteral(5L)))
        }
//        test("primitives support") {
//            buildResults("src/test/resources/primitives").forAll {
//                it.commandResult shouldBe it.expected
//                it.queryResult shouldBe it.expected
//            }
//        }
//
//        test("assignment support") {
//            buildResults("src/test/resources/assignment").forAll {
//                it.commandResult shouldBe it.expected
//                it.queryResult shouldBe it.expected
//            }
//        }
    }
}

File: wander/shared/src/test/scala/dev/ligature/wander/new/WanderSuite.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.wander.`new`

import munit.FunSuite
import munit.IgnoreSuite

class WanderSuite extends FunSuite {
  test("eval empty script") {
    val input = ""
    val expected = "nothing"
    assertEquals(runPrint(input), expected)
  }
  test("eval nothing keyword") {
    val input = "nothing"
    val expected = "nothing"
    assertEquals(runPrint(input), expected)
  }
  test("eval integer literal") {
    val input = "-5"
    val expected = "-5"
    assertEquals(runPrint(input), expected)
  }
  test("eval integers") {
    val input = "-1 0 500 -5000"
    val expected = "-5000"
    assertEquals(runPrint(input), expected)
  }
  test("eval string literal") {
    val input = "\"hello\""
    val expected = "\"hello\""
    assertEquals(runPrint(input), expected)
  }
  test("eval boolean literal") {
    val input = "false"
    val expected = "false"
    assertEquals(runPrint(input), expected)
  }
  test("eval Identifier") {
    val input = "<hello>"
    val expected = "<hello>"
    assertEquals(runPrint(input), expected)
  }
}
